In [1]:
import pandas as pd
import numpy as np
import math
#imports

In [2]:
df=pd.read_csv("csvs\HousingData.csv")
df.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD  TAX  PTRATIO  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900    1  296     15.3   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671    2  242     17.8   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671    2  242     17.8   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622    3  222     18.7   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622    3  222     18.7   

        B  LSTAT  MEDV  
0  396.90   4.98  24.0  
1  396.90   9.14  21.6  
2  392.83   4.03  34.7  
3  394.63   2.94  33.4  
4  396.90    NaN  36.2

In [70]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    int64  
 9   TAX      506 non-null    int64  
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  MEDV     506 non-null    float64
dtypes: float64(12), int64(2)
memory usage: 55.5 KB


In [4]:
df.describe()

CRIM          ZN       INDUS        CHAS         NOX          RM  \
count  486.000000  486.000000  486.000000  486.000000  506.000000  506.000000   
mean     3.611874   11.211934   11.083992    0.069959    0.554695    6.284634   
std      8.720192   23.388876    6.835896    0.255340    0.115878    0.702617   
min      0.006320    0.000000    0.460000    0.000000    0.385000    3.561000   
25%      0.081900    0.000000    5.190000    0.000000    0.449000    5.885500   
50%      0.253715    0.000000    9.690000    0.000000    0.538000    6.208500   
75%      3.560263   12.500000   18.100000    0.000000    0.624000    6.623500   
max     88.976200  100.000000   27.740000    1.000000    0.871000    8.780000   

              AGE         DIS         RAD         TAX     PTRATIO           B  \
count  486.000000  506.000000  506.000000  506.000000  506.000000  506.000000   
mean    68.518519    3.795043    9.549407  408.237154   18.455534  356.674032   
std     27.999513    2.105710    8.707259  168.537116    2.164946   91.294864   
min      2.900000    1.129600    1.000000  187.000000   12.600000    0.320000   
25%     45.175000    2.100175    4.000000  279.000000   17.400000  375.377500   
50%     76.800000    3.207450    5.000000  330.000000   19.050000  391.440000   
75%     93.975000    5.188425   24.000000  666.000000   20.200000  396.225000   
max    100.000000   12.126500   24.000000  711.000000   22.000000  396.900000   

            LSTAT        MEDV  
count  486.000000  506.000000  
mean    12.715432   22.532806  
std      7.155871    9.197104  
min      1.730000    5.000000  
25%      7.125000   17.025000  
50%     11.430000   21.200000  
75%     16.955000   25.000000  
max     37.970000   50.000000

In [5]:
df[["CRIM", "ZN", "INDUS", "CHAS", "AGE", "LSTAT"]]=df[["CRIM", "ZN", "INDUS", "CHAS", "AGE", "LSTAT"]].fillna(0)
#selecting the columns with missing values and replacing them with 0

In [6]:
df["CRIM"].replace(0, df["CRIM"].mean())
df["ZN"].replace(0, df["ZN"].mean())             
df["INDUS"].replace(0, df["INDUS"].mean())
df["AGE"].replace(0, df["AGE"].mean())
df["CHAS"].replace(0, df["CHAS"].mean())
df["LSTAT"].replace(0, df["LSTAT"].mean())

#replacing the 0 set above with the mean values


0       4.980000
1       9.140000
2       4.030000
3       2.940000
4      12.212846
         ...    
501    12.212846
502     9.080000
503     5.640000
504     6.480000
505     7.880000
Name: LSTAT, Length: 506, dtype: float64

In [7]:
df.corr()["MEDV"]

CRIM      -0.384120
ZN         0.362292
INDUS     -0.441371
CHAS       0.183844
NOX       -0.427321
RM         0.695360
AGE       -0.356699
DIS        0.249929
RAD       -0.381626
TAX       -0.468536
PTRATIO   -0.507787
B          0.333461
LSTAT     -0.695405
MEDV       1.000000
Name: MEDV, dtype: float64

# Correlation Notes:

Biggest positive correlation with the value of the home is RM - rooms per dewlling, 
because intuitively more rooms equals  bigger surface area, and in turn more expansive and better

Another big but negative correlation is LSTAT - percentage of lower status population, 
because rich people don't want to live near poor people (as they should)

In [8]:
df_corr=df[["INDUS", "RM", "TAX", "PTRATIO", "LSTAT", "MEDV"]]
df_corr
#selecting the columns with the highest correlation

INDUS     RM  TAX  PTRATIO  LSTAT  MEDV
0     2.31  6.575  296     15.3   4.98  24.0
1     7.07  6.421  242     17.8   9.14  21.6
2     7.07  7.185  242     17.8   4.03  34.7
3     2.18  6.998  222     18.7   2.94  33.4
4     2.18  7.147  222     18.7   0.00  36.2
..     ...    ...  ...      ...    ...   ...
501  11.93  6.593  273     21.0   0.00  22.4
502  11.93  6.120  273     21.0   9.08  20.6
503  11.93  6.976  273     21.0   5.64  23.9
504  11.93  6.794  273     21.0   6.48  22.0
505  11.93  6.030  273     21.0   7.88  11.9

[506 rows x 6 columns]

In [9]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, mean_absolute_error
from sklearn.linear_model import LinearRegression

In [10]:
X_train, X_test, y_train, y_test = train_test_split(df[df.columns[0:13]], df["MEDV"], test_size=0.3, random_state=56)

In [11]:
#training the sklearn model on all data
lr_sklearn=LinearRegression()
lr_sklearn.fit(X_train, y_train)

LinearRegression()

In [12]:
y_pred=lr_sklearn.predict(X_test) #predincting the values, and calculating the errors
print(f"MSE: {mean_squared_error(y_test, y_pred)}")
print(f"MAE: {mean_absolute_error(y_test, y_pred)}")


MSE: 29.098088209233634
MAE: 3.5683147478188766


In [13]:
X_train_corr, X_test_corr, y_train_corr, y_test_corr = train_test_split(df_corr[df_corr.columns[1:5]], df["MEDV"], test_size=0.3, random_state=56)

In [14]:
#training the sklearn model on selected data
lr_sklearn_corr=LinearRegression()
lr_sklearn_corr.fit(X_train_corr, y_train_corr)

LinearRegression()

In [15]:
y_pred_corr=lr_sklearn_corr.predict(X_test_corr) #predincting the values, and calculating the errors
print(f"MSE correlated: {mean_squared_error(y_test, y_pred_corr)}")
print(f"MAE correlated: {mean_absolute_error(y_test, y_pred_corr)}")

MSE correlated: 35.25506777407753
MAE correlated: 3.7535940122955025


|  | SkLearn | SkLearn Correlated |
| --- | --- | --- |
| MSE | 30.5 | 35.25 |
| MAE | 3.65 | 3.75 |

# Observations and Conclusions
The error is pretty big compared to the value we're trying to predict. 
I asume it's because most parameters have a low coralatoins to the end values (only 2 having a bigger correlations than .5), but this is natural because the value of realastate is complex, and driven by many facors, some maybe not included here.

Also, I haven't managed to get the from scratch implementation to work. I belive it's because it'd trying to concatanate arrays of different shapes, but I haven't figured it out.

In [54]:
class LinearRegressionS():
    #Define a Linear Regression class to store our relevant functions in
    def __init__(self):

        self.params__ = None
    def gradientDescent(self, X, y, learning_rate=0.00001, iterations=500, batch_size=16):

        X = np.concatenate([X, np.ones_like(y)], axis = 1)
        rows, cols = X.shape

        X = np.append(X, y, axis=1)
    
        w = np.random.random((cols, 1))
        
        for i in range(iterations):
            np.random.shuffle(X)
            
            y_it = X[:, -1].reshape((rows, 1))
            X_it = X[:, :-1]
           
            for batch in range(math.ceil(rows / batch_size)):
                batch_start = batch * batch_size
                
                x_batch = X_it[batch_start : min(
                batch_start + batch_size, X.shape[0])]
                y_batch = y_it[batch_start : min(
                batch_start + batch_size, X.shape[0])]
                
                w -= learning_rate * np.matmul(x_batch.
                transpose(), (np.matmul(x_batch, w) - y_batch))
                self.params__= w

        return self
    
    def predict(self, X):
        X = np.concatenate([X, np.ones(X.shape[0]).reshape
        ((X.shape[0], 1))], axis=1)
        return np.matmul(X, self.params__)

In [89]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score
from sklearn.preprocessing import LabelEncoder
import matplotlib.pyplot as plt
import numpy as np
 
class Linearregression:

    def __init__(self, learning_rate=0.0001, n_iters=1000):
        self.lr = learning_rate
        self.n_iters = n_iters
        self.weights = None
        self.bias = None

    def fit(self, X, y):
        n_samples, n_features = X.shape

        
        self.weights = np.zeros(n_features)
        self.bias = 0

        for _ in range(self.n_iters):
            y_predicted = np.dot(X, self.weights) + self.bias
           
            dw = (1 / n_samples) * np.dot(X.T, (y_predicted - y))
            db = (1 / n_samples) * np.sum(y_predicted - y)

            self.weights -= self.lr * dw
            self.bias -= self.lr * db


    def predict(self, X):
        y_approximated = np.dot(X, self.weights) + self.bias
        return y_approximated
    

In [123]:
#commnet